Docstubgen-ng

tox-wtf opened the pull request:
Hi, here's a very-much draft-state docstubgen-ng. I'll see what I can do about providing XML-formatted output hopefully soon.

The test suite is currently a mess. I'll make it less bad in the future.

I mostly just wanted to have a public draft for easier feedback.

tox-wtf committed:
add wip test suite

tox-wtf committed:
add wip docstubgen-ng

tox-wtf commented:
I'm thinking I'll add a `build-testdata` make target to fetch and build a package with fewer dependencies than curl in the future. Open to suggestions :)

tox-wtf committed:
Write XML formatting logic, refactor, etc

tox-wtf committed:
Remove old testdata

tox-wtf committed:
Write makefile for docstubgen-ng

tox-wtf committed:
Modify README for docstubgen-ng

tox-wtf committed:
Add scdoc manpage for docstubgen-ng

tox-wtf committed:
Tweak testdata

tox-wtf committed:
Tweak docstubgen-ng library and directory output

tox-wtf committed:
Ignore files for docstubgen-ng

tox-wtf committed:
Add tests for docstubgen-ng

tox-wtf committed:
Remove original docstubgen files

tox-wtf commented:
Pushed today's changes; I'll look at it again once I've slept.

seal331 commented:
Good work! However I'd like to move the old docstubgen to an obviously marked "unsupported old code" folder instead of removing it outright, and those TODOs probably still need to be resolved. My email can be found in the Git log, and I'll handle moving docstubgen myself but you might be forced to resolve a merge conflict.

seal331 commented:
Another thing I forgot: the original docstubgen's checker script had a minor update since you began your rewrite, please merge that in

tox-wtf committed:
Add Seal's email

tox-wtf committed:
Add archival folder for original docstubgen

tox-wtf committed:
Tweak test runner logic for curl workaround

tox-wtf committed:
Refactor docstubgen-ng

tox-wtf committed:
Tweak expected test docstsub for fastfetch

tox-wtf committed:
Tweak test suite runner

tox-wtf committed:
Add format_generic and add to common dirs

tox-wtf commented:
> Another thing I forgot: the original docstubgen's checker script had a minor update since you began your rewrite, please merge that in

I ended up just copying the docstubgen code into an archival folder.

tox-wtf commented:
> Good work! However I'd like to move the old docstubgen to an obviously marked "unsupported old code" folder instead of removing it outright, and those TODOs probably still need to be resolved. My email can be found in the Git log, and I'll handle moving docstubgen myself but you might be forced to resolve a merge conflict.

The remaining TODOs in the main code have been resolved. The test suite could use some extension, but that can occur over time as bugs are discovered.

seal331 commented:
LGTM, except for the fact that I can't seem to find an equivalent of the old docstubgen's binary detect feature? If I'm understanding correctly this code tries to figure out if something is a binary or not by checking if it's in an explicitly marked "bin" folder, which breaks in case binaries are stored in libexec (yes, SLFS documents libexec binaries) or a non-standard location (e. g. CDE's /usr/dt or one of the many historical places the SVR4 tools put their binaries in).

IIRC, what the library documenting option did in the original docstubgen was that it checked for specific extensions and, if they were present, put them into the Libraries list and not the Programs list like the original generate-doc-stub. The binary detect option took that one step further by also checking the magic bytes, which was absolutely necessary if you just pointed it on a DESTDIR or else the programs list would be full of random auxiliary files.

seal331 commented:
I'm also unsure how this code handles libraries in non-standard locations, e. g. the SVR4 tools' /usr/5lib and /usr/ucblib.

tox-wtf commented:
I'll read magic-bytes for binary detection soonish.

zeckma-tech commented:
@seal331 Know you can also review and merge this PR if you so wish. @tox-wtf LMK when you think this PR is ready and I'll take a look.

tox-wtf committed:
Fix a bug related content discovery

tox-wtf committed:
Remove some verbose debug output

tox-wtf committed:
Check for magic bytes and shebangs for binaries

tox-wtf commented:
> I'll read magic-bytes for binary detection soonish.

Magic bytes (and shebangs) are now checked for binaries. This does have the unfortunate consequence of registering scripts shipped in, say, curl's documentation, as installed programs. How do we want to handle that?

tox-wtf commented:
> I'm also unsure how this code handles libraries in non-standard locations, e. g. the SVR4 tools' /usr/5lib and /usr/ucblib.

It just checks against its file extension, so non-standard locations shouldn't mess with library detection.
```bash
# = Determine whether a given file is a library
#
# A library is defined as a (non-symlink) file ending in one of the library
# extensions:
# - a
# - dll
# - jar
# - so*
#
# This function checks file extensions. It does not read magic bytes.
#
# == Returns
# 0 => file is a library
# 1 => file is not a library
is_lib() {
    [[ $1 =~ \.(a$|dll$|jar$|so) ]] &&
        [ ! -L "$1" ]
}
```

tox-wtf commented:
Marked as a draft due to a test failure related to finding programs in documentation I'm not sure how to handle yet.

seal331 commented:
> > I'll read magic-bytes for binary detection soonish.
> 
> Magic bytes (and shebangs) are now checked for binaries. This does have the unfortunate consequence of registering scripts shipped in, say, curl's documentation, as installed programs. How do we want to handle that?

Now this is a first I'm seeing of such a thing... I'd say that if those scripts *aren't* example scripts, and are used to, for example, rebuild some random curl-specific doc index, keep them in the list. Otherwise, explicitly excluding any folders named "doc" from being checked for binaries is acceptable collateral damage. Yes, this is a hack, but I can't think of anything either. The original docstubgen would've put those scripts in the doc folder into the program list too.

tox-wtf committed:
Exclude programs within doc directories

tox-wtf commented:
Testing on iputils, `docstubgen-ng` pulls in many extra directories. I might should use regex to check against common directories instead of the current explicit list.
<img width="520" height="1081" alt="image" src="https://github.com/user-attachments/assets/53b29b74-7acd-4179-98a3-c8839cd7eef1" />

tox-wtf committed:
Rework directory detection

Changes:
- Use regex instead of an explicit list of common directories
- Match all directories rather than just leaf directories

tox-wtf committed:
Update test data for new directory detection scheme

tox-wtf commented:
> Testing on iputils, `docstubgen-ng` pulls in many extra directories. I might should use regex to check against common directories instead of the current explicit list.

This is now fixed:
<img width="541" height="1078" alt="image" src="https://github.com/user-attachments/assets/81e2c4ef-4dab-47f9-b1af-57259a0ca2b9" />

tox-wtf committed:
Add applications directory as a common directory

seal331 commented:
> > Testing on iputils, `docstubgen-ng` pulls in many extra directories. I might should use regex to check against common directories instead of the current explicit list.
> 
> This is now fixed: <img alt="image" width="541" height="1078" src="https://private-user-images.githubusercontent.com/140706894/514801246-81e2c4ef-4dab-47f9-b1af-57259a0ca2b9.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjMzMDc2NTIsIm5iZiI6MTc2MzMwNzM1MiwicGF0aCI6Ii8xNDA3MDY4OTQvNTE0ODAxMjQ2LTgxZTJjNGVmLTRkYWItNDdmOS1iMWFmLTU3MjU5YTBjYTJiOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMTE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTExNlQxNTM1NTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03YWI2ZmVlMzBmYjgyYTAyNTliMWFhZTgwNDIwMWI4M2ZjNTNiMjI2YWQ3NDQ2ZmZlNTI1ZjIwYWZmMjNhZTU0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.21sHXR1bfG6sNhgPuRcPXdFCLIu0mknY744BniyrDL4">

Minor output difference with original docstubgen (the original docstubgen doesn't put every binary on a new line, it instead tries to put as many as possible on one until hitting the 80-character limit) but I don't really mind that, it only makes it easier to adjust binary lists.

I do see something that probably should be fixed though... I have no clue how other editors handle such a situation, but my personal view on listing application-specific folders is "if an application has installed a custom folder, and in it are other custom folders, don't list every single one of those folders but only the one closest to /". @Zeckmathederg, your thoughts?

zeckma-tech commented:
Master directory basically. Like if it installs /usr/include/pkg/{lib1,lib2}, just list /usr/include/pkg.

tox-wtf committed:
Omit child directories

tox-wtf commented:
> Master directory basically. Like if it installs /usr/include/pkg/{lib1,lib2}, just list /usr/include/pkg.

Fixed!

tox-wtf committed:
Add cmake library directory to common directories

tox-wtf committed:
Program detection bugfix

This should avoid flagging dynamic libraries as programs

tox-wtf committed:
Improve documentation comments

tox-wtf committed:
Tweak todo comments in docstub

tox-wtf committed:
Update test data

tox-wtf committed:
Add phony targets to makefile

tox-wtf commented:
I've done a bit of cleanup in the last several commits. `docstubgen-ng` is ready for testing on real packages if y'all wanna start generating new docstubs with it. Any feedback thence would be appreciated!

seal331 commented:
I'm not even sure who's being tortured here, me or the floppy disks I've been trying to save since the 7th... sorry for not being able to look into this PR, and I genuinely have no clue when I'll be able to look at it given my current workload. @Zeckmathederg?

zeckma-tech commented:
> I'm not even sure who's being tortured here, me or the floppy disks I've been trying to save since the 7th... sorry for not being able to look into this PR, and I genuinely have no clue when I'll be able to look at it given my current workload. @Zeckmathederg?

I am planning on adding packages like FluidSynth in SLFS, and will test against this PR for that.

zeckma-tech commented on docstubgen/Makefile:
```diff
@@ -1,51 +1,54 @@
 
-clean:
-	rm -rf $(OBJDIR)
+.PHONY all install install-docstubgen install-man clean lint test testdata
```
For `.PHONY`, you need to add `:` after it, like so:
```Make
.PHONY: all install install-docstubgen install-man clean lint test testdata
```

Otherwise, `make` will complain about a missing separator.

tox-wtf replied:
Somehow managed to miss this when running `make`. Thanks!

tox-wtf committed:
Add missing colon to Makefile

zeckma-tech reviewed (changes requested):
If a program or library has the same name as the package name, add a suffix like `-prog`, `-bin`, or `-lib` for the `varlistentry id`. A decent way to figure out what suffix is needed is by checking the suffix of the package name. If it's `lib`, make the suffix `-lib`. If it doesn't contain `lib`, make the suffix `-bin`. Now differing between `-prog` and `-bin` should be a decision left for the editor writing the page, while `-bin` is the default.

tox-wtf committed:
Append to varlistentry id if needed

tox-wtf committed:
Update test data for varlistentry id tweak

tox-wtf commented:
> If a program or library has the same name as the package name, add a suffix like `-prog`, `-bin`, or `-lib` for the `varlistentry id`.

Fixed. Note that I implicitly used `is_prog()` and `is_lib()` to decide whether to append `-lib` or `-bin`.

zeckma-tech reviewed (approved):
Okay, I think this is good enough to be merged. Thanks for all the hard work on this PR.
